uniffi_bindgen/src/bindings/backend.rs

```diff
@@ -0,0 +1,142 @@
+/// Every `Type` referred to in the `ComponentInterface` should map to a corresponding
+/// `CodeType`.
+///
+/// The mapping may be opaque, but the oracle always knows the answer.
+///
+/// In adddition, the oracle knows how to render identifiers (function names,
+/// class names, variable names etc).
+pub trait CodeOracle {
+    fn find(&self, type_: &TypeIdentifier) -> Box<dyn CodeType>;
+
+    /// Get the idiomatic rendering of a class name (for enums, records, errors, etc).
+    fn class_name(&self, nm: &dyn fmt::Display) -> String;
+
+    /// Get the idiomatic rendering of a function name.
+    fn fn_name(&self, nm: &dyn fmt::Display) -> String;
+
+    /// Get the idiomatic rendering of a variable name.
+    fn var_name(&self, nm: &dyn fmt::Display) -> String;
+
+    /// Get the idiomatic rendering of an individual enum variant.
+    fn enum_variant_name(&self, nm: &dyn fmt::Display) -> String;
+
+    /// Get the idiomatic rendering of a module
+    fn mod_name(&self, nm: &dyn fmt::Display) -> String;
+
+    /// Get the idiomatic rendering of an exception name
+    ///
+    /// This replaces "Error" at the end of the name with "Exception".  Rust code typically uses
+    /// "Error" for any type of error but in the Java world, "Error" means a non-recoverable error
+    /// and is distinguished from an "Exception".
+    fn error_name(&self, nm: &dyn fmt::Display) -> String;
+
+    fn ffi_type_label(&self, ffi_type: &FFIType) -> String;
+}
+
+/// A Trait to emit foreign language code to handle referenced types.
+/// A type which is specified in the UDL (i.e. a member of the component interface)
+/// will have a `CodeDeclaration` as well, but for types used e.g. primitive types, Strings, etc
+/// only a `CodeType` is needed.
+pub trait CodeType {
+    /// The language specific label used to reference this type. This will be used in
+    /// method signatures and property declarations.
+    fn type_label(&self, oracle: &dyn CodeOracle) -> String;
+
+    /// A representation of this type label that can be used as part of another
+    /// identifier. e.g. `read_foo()`, or `FooInternals`.
+    ///
+    /// This is especially useful when creating specialized objects or methods to deal
+    /// with this type only.
+    fn canonical_name(&self, oracle: &dyn CodeOracle) -> String {
+        self.type_label(oracle)
+    }
+
+    /// A representation of the given literal for this type.
+    /// N.B. `Literal` is aliased from `interface::Literal`, so may not be whole suited to this task.
+    fn literal(&self, oracle: &dyn CodeOracle, literal: &Literal) -> String;
+
+    /// An expression for lowering a value into something we can pass over the FFI.
+    ///
+    /// N.B. This should align with the `helper_code` generated by this `CodeType`.
+    fn lower(&self, _oracle: &dyn CodeOracle, _nm: &dyn fmt::Display) -> String {
+        panic!("lower not implemented")
+    }
```

Comment:
I added default implementations, because we don't need to implement these for Kotlin anymore.  This makes me want to split up `CodeType` into a language-agnostic version and language-specific version.